File: TurnBased-Mess-main/TurnBased-Mess-main/TurnBased-Mess/ProyectoParejasPOO/ProyectoParejasPOO/TurnManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ProyectoParejasPOO
{
    public class TurnManager
    {
        public List<Character> allUnits = new List<Character>();
        public Queue<Character> turnOrder = new Queue<Character>();
        public int totalTurns = 0;

        public void CalculateOrder()
        {
            turnOrder.Clear();

            List<Character> ordered = allUnits
                .Where(unit => unit != null && unit.isAlive)
                .OrderByDescending(unit => unit.spd)
                .ToList();

            foreach (var unit in ordered)
                turnOrder.Enqueue(unit);
        }

        public Character NextTurn()
        {
            if (turnOrder.Count == 0)
                StartRound();

            RemoveDeadCharacters();

            Character unit = turnOrder.Dequeue();
            totalTurns++;
            BattleUI.ShowTurnStart(unit);
            return unit;
        }

        public void RemoveDeadCharacters()
        {
            allUnits = allUnits.Where(unit => unit != null && unit.isAlive).ToList();

            if (turnOrder.Count > 0)
            {
                var remaining = turnOrder.Where(unit => unit != null && unit.isAlive).ToList();
                turnOrder.Clear();
                foreach (var unit in remaining)
                    turnOrder.Enqueue(unit);
            }
        }

        public void StartRound()
        {
            RemoveDeadCharacters();
            CalculateOrder();
        }

        public void AddUnit(Character c)
        {
            if (c == null)
                return;

            if (!allUnits.Contains(c))
                allUnits.Add(c);
        }

        public void ClearUnits()
        {
            allUnits.Clear();
            turnOrder.Clear();
            totalTurns = 0;
        }
    }
}

File: Java/ProyectoParejasPOOJava/src/TurnManager.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class TurnManager {

    public List<Character> allUnits  = new ArrayList<>();
    public Queue<Character> turnOrder = new LinkedList<>();
    public int totalTurns = 0;

    public void calculateOrder() {
        turnOrder.clear();
        List<Character> alive = new ArrayList<>();
        for (Character u : allUnits)
            if (u != null && u.isAlive) alive.add(u);

        // Sort by spd descending
        alive.sort((a, b) -> b.spd - a.spd);

        turnOrder.addAll(alive);
    }

    public Character nextTurn() {
        if (turnOrder.isEmpty()) startRound();
        removeDeadCharacters();
        Character unit = turnOrder.poll();
        totalTurns++;
        BattleUI.showTurnStart(unit);
        return unit;
    }

    public void removeDeadCharacters() {
        allUnits.removeIf(u -> u == null || !u.isAlive);

        List<Character> remaining = new ArrayList<>();
        for (Character u : turnOrder)
            if (u != null && u.isAlive) remaining.add(u);
        turnOrder.clear();
        turnOrder.addAll(remaining);
    }

    public void startRound() {
        removeDeadCharacters();
        calculateOrder();
    }

    public void addUnit(Character c) {
        if (c == null) return;
        if (!allUnits.contains(c)) allUnits.add(c);
    }

    public void clearUnits() {
        allUnits.clear();
        turnOrder.clear();
        totalTurns = 0;
    }
}
